bootstrap/src/lib/cache/rateLimit.ts

```diff
@@ -0,0 +1,158 @@
+import { logger } from '@chainlink/external-adapter'
+
+type RateLimitOptions = {
+  participantMaxAge: number
+  groupId: string
+  totalCapacity: number
+  maxAgeLimits: {
+    min: number
+    max: number
+  }
+}
+
+export interface RateLimit {
+  isEnabled: boolean
+  getParticipantMaxAge: ((participantId: string) => number) | (() => void)
+  incrementParticipantHeartbeat: ((participantId: string, cost?: number) => void) | (() => void)
+  newParticipant: ((participantId: string, cost?: number) => void) | (() => void)
+}
+
+type Participant = {
+  cost: number
+  heartbeat: number
+  lastSeen: number
+}
+
+let participants: Map<string, Participant> = new Map()
+
+const makeParticipants = (participantMaxAge: number) => {
+  const DEFAULT_COST = 1
+
+  const getParticipant = (participantId: string): Participant | undefined => {
+    return participants.get(participantId)
+  }
+
+  const createParticipant = (participantId: string, cost?: number): Participant => {
+    let currentParticipant = getParticipant(participantId)
+    // If the participant exists, we don't create a new one. It updates the cost in case it exists
+    if (currentParticipant) {
+      if (cost && cost !== currentParticipant.cost) {
+        currentParticipant = {
+          cost: cost,
+          heartbeat: currentParticipant.heartbeat,
+          lastSeen: currentParticipant.lastSeen,
+        }
+        participants.set(participantId, currentParticipant)
+      }
+      return currentParticipant
+    }
+
+    const participant = { cost: cost || DEFAULT_COST, heartbeat: 0, lastSeen: Date.now() }
+    participants.set(participantId, participant)
+    return participant
+  }
+
+  const increaseParticipantHeartbeat = (participantId: string) => {
```

Comment:
When is the heartbeat decremented?

We should be keeping a sliding window of requests so we can answer questions like:

1. How many requests of type did we observe in the last 24h?
2. How many in the last hour?
3. How many in the last minute?